presto-plan-checker-providers/src/main/java/com/facebook/presto/plancheckerproviders/nativechecker/NativePlanChecker.java

```diff
@@ -0,0 +1,182 @@
+        this.config = requireNonNull(config, "config is null");
+        this.httpClient = new OkHttpClient.Builder().build();
+    }
+
+    private URL getPlanValidateUrl()
+    {
+        Set<Node> nodes = nodeManager.getAllNodes();
+        Node sidecarNode = nodes.stream().filter(Node::isCoordinatorSidecar).findFirst()
+                .orElseThrow(() -> new PrestoException(NOT_FOUND, "could not find native sidecar node"));
+        try {
+            // endpoint to retrieve session properties from native worker
+            return new URL(sidecarNode.getHttpUri().toString() + PLAN_CONVERSION_ENDPOINT);
+        }
+        catch (MalformedURLException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public void validate(PlanNode planNode, WarningCollector warningCollector)
+    {
+        // NO-OP, only validating fragments
+    }
+
+    @Override
+    public void validateFragment(SimplePlanFragment planFragment, WarningCollector warningCollector)
+    {
+        try {
+            if (!planFragment.getPartitioning().isCoordinatorOnly() && !isInternalSystemConnector(planFragment.getRoot())) {
+                runValidation(planFragment);
+            }
+            else if (LOG.isDebugEnabled()) {
+                LOG.debug("Skipping Native Plan Validation for plan fragment id: %s", planFragment.getId());
+            }
+        }
+        catch (final PrestoException e) {
+            if (config.isQueryFailOnError()) {
+                throw e;
+            }
+        }
+        catch (final Exception e) {
+            LOG.error(e, "Failed to run native plan validation");
+            if (config.isQueryFailOnError()) {
+                throw new IllegalStateException("Failed to run native plan validation", e);
+            }
+        }
+    }
+
+    private boolean isInternalSystemConnector(PlanNode planNode)
+    {
+        try {
+            return planNode.accept(new CheckInternalVisitor(), null);
+        }
+        catch (final IllegalArgumentException e) {
+            // An InternalPlanNode will not allow a `PlanVisitor`, allow it to pass
+            return false;
+        }
+    }
+
+    private void runValidation(SimplePlanFragment planFragment)
+            throws IOException
+    {
+        LOG.debug("Starting native plan validation for plan fragment id: %s", planFragment.getId());
+
+        String requestBodyJson = planFragmentJsonCodec.toJson(planFragment);
+        LOG.debug("Native validation for plan fragment: %s", requestBodyJson);
+
+        Request request = buildRequest(requestBodyJson);
+
+        try (Response response = httpClient.newCall(request).execute()) {
+            if (!response.isSuccessful()) {
+                String responseBody = response.body() != null ? response.body().string() : "{}";
+                LOG.error("Native plan checker failed with code: %d, response: %s", response.code(), responseBody);
```

Comment:
That is fine, I just want to make sure it's clear that if failed due to native plan checker, so if the exception message contains that and is logged, then that should be fine.

Reply:
If it doesn't, it should be changed so that it does.  It's much more helpful to have the failure reason show up in the query error than to have to scour the coordinator logs for any relevant error messages.